### PR TITLE
[Train] Micro-batch scheduling on rollout side + variable global batch size (port slime #1926+#1930+#1933)

### PR DESCRIPTION
**Pull request by aoshen02**
Ports the coupled THUDM/slime chain **#1926 → #1930 → #1933** (one variable-global-batch feature line, three sequential PRs) into vime as **one PR** — they touch the same files interdependently (`dp_schedule.py`, `actor.py`, `data.py`, `model.py`, `loss.py`, `ray/rollout.py`), so splitting would mean reconciling the same diverged files three times and a merge-order hazard. One PR = one CI run.

## What
- **#1926** — move the per-step micro-batch schedule from the actor (train) side to `_split_train_data_by_dp` on the **rollout** side, computed once over the full global batch. New **`vime/utils/dp_schedule.py`** (`build_dp_schedule`, `compute_dynamic_global_batch_size`) holds the Ray/engine-free logic so it's unit-testable. Drops the DP-wide `all_reduce(num_microbatches, MAX)`; `get_data_iterator` becomes a thin consumer reading `num_microbatches` / `micro_batch_indices` out of `rollout_data`.
- **#1930 + #1933 [1/N, 2/N]** — variable global batch size: per-step split by rollout id so each rollout yields a fixed training-step count regardless of how many samples it produced; CP-aware loss / metric handling.

## vime reconciliation
- vime already carried `--use-dynamic-global-batch-size` / `--disable-rollout-trim-samples` (args + an inline `_compute_dynamic_global_batch_size`) from an earlier partial sync; #1926 removes that inline method in favor of `dp_schedule.compute_dynamic_global_batch_size` — applied cleanly.
- `_split_train_data_by_dp` **keeps vime's diverged extra keys** (`rollout_routed_experts`, `teacher_log_probs`, `dynamic_global_batch_size`).
- `train_parallel_config` extended with `cp_size` / `vpp_size` / `microbatch_group_size_per_vp_stage` (set in `actor.py:92`, read by `build_dp_schedule`).
- Two rollout.py import hunks hand-reconciled (vime's vLLM seam — no `SGLangEngine` import).
- Patch applied via path+import rewrite (`slime/`→`vime/`, `slime.`→`vime.`); `/root/slime` mount and `SLIME_TEST_*` envs preserved.

## Tests / CI
Adds the chain's **CPU** tests to the always-on cpu job: `test_dp_schedule`, `test_cp_utils`, `test_loss_cp_invariance`, `test_metric_report`, `test_metric_report_dist` (+ `tests/_cp_dist_helpers.py`). j2 updated + yml re-rendered.

## Validation
- py_compile clean across all 16 touched files; pinned pre-commit (ruff/isort/black) clean.
- cpu CI (1916 image): see PR comment once the run completes.
- ⚠️ **This changes training micro-batch counts / bin-packing (numerics) → needs GPU validation before merge.** GPU CI is currently off per maintainer instruction; cpu validation covers the shipped CPU tests only.

Upstream: THUDM/slime#1926, #1930, #1933. Part of the slime→vime sync tracked in #107 (chain registered there; not ticking boxes pending sign-off).

### CODE REVIEW

**Comment by aoshen02** [comment]
**cpu CI — all green** (fresh `--rm` container, latest cumulative image `inferactinc/public:vime-vllm-cu129-sync1916`, `--gpus all`):

| group | result |
|---|---|
| `test_dp_schedule` | 8 passed |
| `test_cp_utils` | 5 passed |
| `test_loss_cp_invariance` | 6 passed |
| `test_metric_report` | 19 passed |
| `test_metric_report_dist` | 27 passed |
| `test_megatron_argument_validation` | 7 passed |
| plugin_contracts (rollout/runtime-hook/path-loading/generate) | 6 / 10 / 14 / 4 = 34 passed |
| unit suite (per-file parallel) | 0 failed of 11 |

The five tests the chain ships (dp_schedule scheduling, CP-utils, CP loss-invariance, metric reporting incl. distributed) exercise the core of the refactor on CPU. ⚠️ Still **needs GPU validation before merge** — this changes training micro-batch counts / bin-packing; GPU CI is off per maintainer instruction.

File: vime/backends/megatron_utils/cp_utils.py [hunk 1]
```diff
@@ -120,6 +136,114 @@ def sum_of_token(x: torch.Tensor) -> torch.Tensor:
...
+            first = values[0]
+            if isinstance(first, tuple) and len(first) == 2:
+                total_sum = sum(v[0] for v in values)
+                total_count = sum(v[1] for v in values)
+                reduced[key] = total_sum / total_count if total_count else 0.0
+            else:
+                reduced[key] = sum(values) / dp_size
+        return reduced
+    dist.gather_object(log_dict, None, dst=dp_src_rank, group=dp_group)
+    return None
```
[the remark this answers is not in the document]
Comment (aoshen02) [reply vime/backends/megatron_utils/cp_utils.py:244]: This is a false positive — keeping as-is.

For `dist.gather_object`, **`dst` is the *global* rank** (PyTorch docstring: *"Destination rank on global process group (regardless of `group` argument)"*); the group-relative parameter is `group_dst`. The guard here is `if dist.get_rank() == dp_src_rank`, and `dist.get_rank()` returns the global rank — so `dst=dp_src_rank` is in the same (global) coordinate system and is correct.

Switching to `dst=0` would actually *introduce* the bug you describe: on multi-node, the DP group's first rank is a non-zero global rank, so gathering to global rank 0 while the receiving branch is `dp_src_rank` would mismatch. This also matches slime upstream verbatim. No change.

File: vime/utils/dp_schedule.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+    for step_i in range(num_steps):
+        step_rollouts = rollout_ids[step_i * global_batch_size : (step_i + 1) * global_batch_size]
+        sample_indices = [pos for rid in step_rollouts for pos in rollout_id_to_samples[rid]]
+        step_lengths = [total_lengths[i] for i in sample_indices]
+        global_batch_sizes.append(global_batch_size)
+        assert len(sample_indices) >= dp_size, (
+            f"step {step_i}: {len(sample_indices)} samples < dp_size {dp_size}; "
+            f"each step needs at least one sample per rank."
+        )
```
[the remark this answers is not in the document]
Comment (aoshen02) [reply vime/utils/dp_schedule.py:138]: Valid but minor — keeping `>= dp_size` for now. The downstream alignment assert (`target_K`/`align_to`) already rejects the VPP-with-too-few-samples case, just later in the call. Holding this byte-aligned with slime upstream to ease future syncs; noted as a follow-up if we want an earlier/clearer error message.

**Comment by aoshen02** [comment]
Closing for now (slime→vime sync being restructured). Branch is preserved — can reopen anytime.
